### PR TITLE
fix: ws errors should conform to jsonrpc spec

### LINKED ISSUE #2362: WS server limiter close connection at max limit wihtout sending the reason back to clients
### Problem

WS server limiter close connection at max limit wihtout sending the reason back to clients

### Solution

add ability to send back closing reason to clients

### Alternatives

_No response_

### PR DESCRIPTION
**Pull request by ted-yeh-vc**
**Description**:

This PR fixes the error returned from websockets being closed to be conformant to json rpc spec re: errors.
https://www.jsonrpc.org/specification#error_object
**Related issue(s)**:

Fixes #2362

**Notes for reviewer**:

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [ ] Tested (unit, integration, etc.)

### CODE REVIEW

**Comment by ted-yeh-vc** [comment]
I could be in the wrong, but it seems unnecessary to push an error to the client before closing the connection. The connection closing has its own code and message. Was there a particular reason the separate message is being sent?

**Comment by quiet-node** [comment]
> I could be in the wrong, but it seems unnecessary to push an error to the client before closing the connection. The connection closing has its own code and message. Was there a particular reason the separate message is being sent?

Hi @ted-yeh-vc thanks for poiniting it out! You're right the closing connection has its own code and message, when I pushed in the update I thought it could make it a little more informative for some users by informing them the error before closing the connection. Just a little slight touch on the UX. I will discuss with the team and get back to you later!

File: packages/ws-server/src/metrics/connectionLimiter.ts [hunk 1]
```diff
@@ -133,7 +133,10 @@ export default class ConnectionLimiter {
       ctx.websocket.send(
         JSON.stringify({
           jsonrpc: '2.0',
-          error: `Closing current connection due to exceeded maximum connections (max_con=${MAX_CONNECTION_LIMIT})`,
+          error: {
+            code: -32603,
+            message: `Closing current connection due to exceeded maximum connections (max_con=${MAX_CONNECTION_LIMIT})`,
+          },
```
Comment (quiet-node) [inline packages/ws-server/src/metrics/connectionLimiter.ts:139]:
```suggestion
          error: {
            code: CONNECTION_LIMIT_EXCEEDED.code,
            message: CONNECTION_LIMIT_EXCEEDED.message,
            data: {
              message: CONNECTION_LIMIT_EXCEEDED.message,
              max_connection: MAX_CONNECTION_LIMIT,
            },
          },
```

File: packages/ws-server/src/metrics/connectionLimiter.ts [hunk 2]
```diff
@@ -152,7 +155,10 @@ export default class ConnectionLimiter {
       ctx.websocket.send(
         JSON.stringify({
           jsonrpc: '2.0',
-          error: `Closing current connection due to exceeded maximum connections from a single IP: address ${ip} - ${this.clientIps[ip]} connections. (max_con=${MAX_CONNECTION_LIMIT_PER_IP})`,
+          error: {
+            code: -32603,
+            message: `Closing current connection due to exceeded maximum connections from a single IP: address ${ip} - ${this.clientIps[ip]} connections. (max_con=${MAX_CONNECTION_LIMIT_PER_IP})`,
+          },
```
Comment (quiet-node) [inline packages/ws-server/src/metrics/connectionLimiter.ts:161]:
```suggestion
          error: {
            code: CONNECTION_IP_LIMIT_EXCEEDED.code,
            message: CONNECTION_IP_LIMIT_EXCEEDED.message,
            data: {
              message: CONNECTION_IP_LIMIT_EXCEEDED.message,
              max_connection: MAX_CONNECTION_LIMIT_PER_IP,
            },
          },
```

File: packages/ws-server/src/metrics/connectionLimiter.ts [hunk 3]
```diff
@@ -188,7 +194,10 @@ export default class ConnectionLimiter {
           websocket.send(
             JSON.stringify({
               jsonrpc: '2.0',
-              error: `Closing current connection due to reaching TTL (${maxInactivityTTL}ms)`,
+              error: {
+                code: -32603,
+                message: `Closing current connection due to reaching TTL (${maxInactivityTTL}ms)`,
+              },
```
Comment (quiet-node) [inline packages/ws-server/src/metrics/connectionLimiter.ts:200]:
```suggestion
              error: {
                code: TTL_EXPIRED.code,
                message: TTL_EXPIRED.message,
                data: {
                  message: TTL_EXPIRED.message,
                  max_inactivity_TTL: maxInactivityTTL,
                },
              },
```

**Review by quiet-node — CHANGES_REQUESTED** [review]
So, I discussed with my team and we decided to keep the current layout with the extra piece of message sent back to the client. Also, we aggreed that your approach to make the error conform to jsonrpc spec is necessary, and we want the errors to have a consistent structure. Please find the change suggestion below for the new structure. Thanks a lot for the time and effort!

**Comment by ted-yeh-vc** [comment]
Fantastic. Thanks for the quick review. Always happy to help!

**Review by quiet-node — APPROVED** [review]
LGTM

**Comment by quiet-node** [comment]
@ted-yeh-vc looks like the first commit didn't get verified. Please fix it for the PR to be mergable.

**Review by Nana-EC — APPROVED** [review]
LGTM

**Pull request merged**